designs/embed_eksd_manifest.md

```diff
@@ -0,0 +1,72 @@
+It would be ideal to have this manifest and its CRDs embedded in the cluster to allow the controller to operate without the requirement to make that external call.
+This will eventually be needed for air-gapped support.
+
+### Goals and Objectives
+
+As an EKS Anywhere user:
+
+* I want the EKSA controller to be able to operate without internet access
+* I want EKSD components to be supported during upgrade
+
+## Overview of Solution
+
+The EKS Distro manifest and the release CRD will be applied directly to the cluster.
+This will allow the controller to grab the EKSD release resources from within the cluster instead of reading in the EKSD manifest from the internet when fetching the applied spec.
+
+### Solution Details
+
+With this feature, the EKSD manifest & its CRD will be applied directly to the cluster.
+We will do this similarly to how we apply the EKSA components to the cluster, introducing a new task to the create workflow called `InstallEksdComponentsTask`.
+We currently don’t pull in the EKSD release CRD in the bundle, so I propose introducing a new field `crdUrl` in the EKSD Bundle to refer to the file in S3 that stores this CRD.
+
+```
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: Bundles
+metadata:
+  creationTimestamp: "2022-02-01T16:41:19Z"
+spec:
+  cliMaxVersion: v0.0.0
+  cliMinVersion: v0.0.0
+  number: 1
+  versionsBundles:
+    eksD:
+      channel: 1-21
+      crdUrl: https://distro.eks.amazonaws.com/crds/releases.distro.eks.amazonaws.com-v1alpha1.yaml
+      gitCommit: |
+        fcddf59a516102e20f75b6c672e3aa74cf2da877
+      kindNode:
+        arch:
+```
+
+From here, we can use `kubectl apply` to apply the content from the EKSD release URL and the CRD URL to the cluster.
+
+We will need a way to refer to the EKSD release object on the cluster when calling it from the controller.
+One way to achieve this is by adding a field `EkdsReleaseRef` to the cluster status.
+```
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: Cluster
+metadata:
+  name: test_cluster
+spec:
+  controlPlaneConfiguration:
+    ...
+status:
+  eksdReleaseRef:
+    name: "name of resource" 
+    url: "url we downloaded from"
+```
+
+From here, we can call `kubectl get release` using the `eksdReleaseRef.Name` to access the EKSD release manifest from the controller.
+
+### Upgrading EKSD components
+
+Now that the EKSD components are embedded in the cluster, we have to ensure that those components are supported during upgrade.
+We will update the upgrader in cluster manager to not only check for updates to EKSA, but it will also check if updates are needed for the EKSD objects on the cluster.
+We will have a method `EksdChangeDiff` to check both the EKSD version in the bundle and the Kubernetes version in the cluster config to decide if we need to update the EKSD manifest objects that are applied to the cluster.
```

Comment:
nit: it should be enough to check the eks-d version, since a different kubernetes minor version guarantees a new eks-d release.

Reply:
I was thinking to add this in the case where the user, for example, creates a cluster with kubeVersion 1.20 & upgrades the cluster to use 1.21